app/src/main/kotlin/compiler/analysis/BuiltinFunctions.kt

```diff
@@ -1,96 +1,32 @@
-                        analyzeExpression(expression.resultWhenFalse)
-                    }
-
-                    is Expression.Variable -> {}
-                    is Expression.BooleanLiteral -> {}
-                    is Expression.NumberLiteral -> {}
-                    is Expression.UnitLiteral -> {}
-                    null -> {}
-                }
-            }
-
-            fun analyzeVariable(node: Variable) { node.value?.let { analyzeExpression(it) } }
-            fun analyzeParameter(node: Function.Parameter) { node.defaultValue?.let { analyzeExpression(it) } }
-            fun analyzeFunction(node: Function, analyzeStatement: (Statement) -> Unit) {
-                node.parameters.forEach { analyzeParameter(it) }
-                node.body.forEach { analyzeStatement(it) }
+    fun addBuiltinFunctions(program: Program): Program {
+        val globalNames = program.globals.map {
+            when (it) {
+                is Program.Global.FunctionDefinition -> it.function.name
+                is Program.Global.VariableDefinition -> it.variable.name
             }
-
-            fun analyzeStatement(statement: Statement) {
-                when (statement) {
-                    is Statement.Evaluation -> { analyzeExpression(statement.expression) }
-                    is Statement.Assignment -> { analyzeExpression(statement.value) }
-                    is Statement.Block -> { statement.block.forEach { analyzeStatement(it) } }
-
-                    is Statement.Conditional -> {
-                        analyzeExpression(statement.condition)
-                        statement.actionWhenTrue.forEach { analyzeStatement(it) }
-                        statement.actionWhenFalse?.forEach { analyzeStatement(it) }
-                    }
-
-                    is Statement.Loop -> {
-                        analyzeExpression(statement.condition)
-                        statement.action.forEach { analyzeStatement(it) }
-                    }
-
-                    is Statement.FunctionReturn -> { analyzeExpression(statement.value) }
-                    is Statement.VariableDefinition -> { analyzeVariable(statement.variable) }
-                    is Statement.FunctionDefinition -> { analyzeFunction(statement.function, ::analyzeStatement) }
-                    is Statement.LoopBreak -> {}
-                    is Statement.LoopContinuation -> {}
-                }
-            }
-
-            when (global) {
-                is Program.Global.VariableDefinition -> { analyzeVariable(global.variable) }
-                is Program.Global.FunctionDefinition -> { analyzeFunction(global.function, ::analyzeStatement) }
-            }
-        }
-
-        program.globals.forEach { analyzeGlobal(it) }
-        return result
+        }.toSet()
+        return Program(builtinFunctions.filter { it.key !in globalNames }.map { Program.Global.FunctionDefinition(it.value) } + program.globals)
```

Comment:
Why allow overriding built-in functions?

Reply:
Do we assume that?
I treat them as functions declared at level -1, which means that global functions (at level 0) can override them.

Reply:
If we were to add some module system, then it would make sense for the standard library to be some implicitly imported module, but the import statement is usually understood to introduce symbols into the global scope. The existence of "level -1" is a bit at odds with the meaning of the word "global" in "global scope".

Reply:
(In the sense that the global scope is usually *the uppermost* scope.)

Reply:
But I suppose this consistency is not that much important - we can make an exception for the "implicitly imported" symbols just for the sake of convenience. You can leave this is as it is if the arguments above are not convincing.

Reply:
Well, I treated "standard library" and "builtin functions" as two different things: "Standard library" could potentially have more functions which are not imported by default and "Builtin functions" as subset of standard library and these function are imported by the compiler (potentially in different way than "global import" you mentioned).
But if it significantly changes our plans then I don't insist.

Reply:
We don't really have any plans about these matters yet, only assumptions. And this distinction between "standard library" and "built-in functions" seems quite reasonable, so let's leave this as it is.